### ISSUE TITLE
git_diff_get_stats fails when LC_ALL=""

### ISSUE THREAD

**Issue by rgburke** [libgit2/libgit2#4528]
Generating diffs for some commits fails with the error `Invalid collation character` when `setlocale(LC_ALL, "");` has been called. A similar issue nodegit/nodegit#1097 suggests that `regcomp` called from `git_diff_driver_builtin` is the source of the error.

### Reproduction steps
Run the following program on the Linux kernel source tree (https://github.com/torvalds/linux):

```c
#include <locale.h>
#include <stdio.h>
#include <git2.h>

int main(int argc, char *argv[]) {
	setlocale(LC_ALL, "");

        git_libgit2_init();

        git_repository *repo;
        git_repository_open(&repo, ".git");

	git_object *obj = NULL;
	git_revparse_single(&obj, repo, "f8de483e7440b0d23ce6372b3ef8358841c8827b");

	git_commit *commit = NULL;
	git_commit_lookup(&commit, repo, git_object_id(obj));

	git_commit *parent = NULL;
	git_commit_parent(&parent, commit, 0);

	git_tree *commit_tree = NULL, *parent_tree = NULL;
	git_commit_tree(&commit_tree, commit);
	git_commit_tree(&parent_tree, parent);

	git_diff *diff = NULL;
	git_diff_tree_to_tree(&diff, repo, parent_tree, commit_tree, NULL);

	git_diff_stats *stats;
	int error = git_diff_get_stats(&stats, diff);

	if (error < 0) {
		fprintf(stderr, "ERROR %d: Could not get diff stats: %s\n", error, giterr_last()->message);
	}

        return 0;
}
```

### Expected behavior
Running this program should produce no output. When `setlocale(LC_ALL, "");` is commented out it produces no output as expected:

```
richard@DevNew:~/code/linux$ ./diff 
richard@DevNew:~/code/linux$ 
```

### Actual behavior
`git_diff_get_stats` fails with the error below:

```
richard@DevNew:~/code/linux$ ./diff 
ERROR -1: Could not get diff stats: Invalid collation character
richard@DevNew:~/code/linux$ 
```

### Version of libgit2 (release number or SHA1)
Using master (408b16c1b06526f509fdb8997818d648d4d1ea1e)

### Operating system(s) tested
Tested on Ubuntu 17.10:
```
Linux DevNew 4.13.0-32-generic #35-Ubuntu SMP Thu Jan 25 09:13:46 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by sebastiencs** [comment]
I also have this error with `git_diff_foreach` on the emacs repository (git://git.savannah.gnu.org/emacs.git)

**Comment by pks-t** [comment]
If this really is `regcomp` failing then there unfortunately is nothing we can do about it. libgit2 cannot set the locale with `setlocale`, as that is global to the complete process and the caller may want to set specific locales on purpose. Due to that, the caller needs to be responsible for setting up the locale environment correctly. And as git usually expects UTF8 encoded data, I'd wager that the locale set should be something which is able to handle UTF8 correctly.

Do these issues persist in case where the language is set to some UTF8 language?

**Comment by sebastiencs** [comment]
In my case, it always fails, no matter which locale is set (including `en_US.UTF-8` and `fr_FR.UTF-8`)

**Comment by rsrchboy** [comment]
I'm seeing the same issue; in my case, if I set `LC_COLLATE=C` things work as expected.  (Otherwise everything is `en_US.utf8`)

**Comment by pks-t** [comment]
I'm able to reproduce the error. The issue is that we use ranges ("[a-z0-9]"), which has different results based on the collation. I'll come up with a PR to fix this

**Comment by pks-t** [comment]
See #4560. Input would be highly appreciated, as the issue is non-trivial to solve and has different drawbacks based on what solution is being used in the end

**Comment by rgburke** [comment]
Thanks for looking into this @pks-t. I can confirm that #4560 fixes the issue on the Linux repository, I no longer get the `Invalid collation character` error.

**Comment by rgburke** [comment]
@pks-t Is there any prosptect of #4560 getting merged? I can understand it's not ideal to diverge from git's regexes, however it does resolve the issue.

**Comment by pks-t** [comment]
Sorry this is taking so long. Thing is that I myself am not totally sure whether #4560 regresses platforms with weird locales, and I cannot test this either as my system cannot load different languages at all. I'll try to take another stab at that issue and see whether I'm able to come up with a safer solution.

**Comment by rgburke** [comment]
Thanks for taking the time to look into this, I really appreciate it. I can see from the PR that this is a difficult problem to solve.

**Comment by ethomson** [comment]
Fixed via #4935
